runtime/near-vm-runner/src/wasmer2_runner.rs

```diff
@@ -266,26 +272,106 @@ impl Wasmer2VM {
+                        unsafe {
+                            // (UN-)SAFETY: the `serialized_module` must have been produced by a prior call to
+                            // `serialize`.
+                            //
+                            // In practice this is not necessarily true. One could have forgotten to change the
+                            // cache key when upgrading the version of the wasmer library or the database could
+                            // have had its data corrupted while at rest.
+                            //
+                            // There should definitely be some validation in wasmer to ensure we load what we think
+                            // we load.
+                            let executable =
+                                UniversalExecutableRef::deserialize(&serialized_module)
+                                    .map_err(|_| CacheError::DeserializationError)?;
+                            let artifact = self
+                                .engine
+                                .load_universal_executable_ref(&executable)
+                                .map(Arc::new)
+                                .map_err(|_| CacheError::DeserializationError)?;
+                            Some(artifact)
+                        }
+                    }
+                };
+
+                let artifact: VMArtifact = match stored_artifact {
+                    Some(it) => it,
+                    None => {
+                        let executable_or_error = self.compile_uncached(code);
+                        if let Some(cache) = cache {
+                            let record = match &executable_or_error {
+                                Ok(executable) => {
+                                    let code = executable.serialize().map_err(|_e| {
+                                        CacheError::SerializationError { hash: key.0 }
+                                    })?;
+                                    CompiledContract::Code(code)
+                                }
+                                Err(err) => CompiledContract::CompileModuleError(err.clone()),
+                            };
+                            cache
+                                .put(&key, record)
+                                .map_err(|_: io::Error| CacheError::WriteError)?;
+                        }
+                        match executable_or_error {
+                            Ok(executable) => self
+                                .engine
+                                .load_universal_executable(&executable)
+                                .map(Arc::new)
+                                .map_err(|_| CacheError::DeserializationError)?,
+                            Err(it) => return Ok(Err(it)),
+                        }
+                    }
+                };
+
+                Ok(Ok(artifact))
+            };
+
+        #[cfg(feature = "no_cache")]
+        return compile_or_load_from_store();
```

Comment:
Need to test whether compilation with `no_cache` still works – I don’t think this name exists?